wlog/auditlog/audit3log/internal/categories/category.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2025 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package categories
+
+import (
+	"fmt"
+	"reflect"
+
+	werror "github.com/palantir/witchcraft-go-error"
+	commonv2 "github.com/palantir/witchcraft-go-logging/conjure/foundry/audit/api/common/v2"
+	"github.com/palantir/witchcraft-go-logging/conjure/witchcraft/api/logging"
+)
+
+// Converts the provided data to a slice of v2.Resource. Based on the Java "promote" method.
+func toResource(data any) ([]commonv2.Resource, error) {
```

Comment:
This file and all functions are based very closely on implementation/logic in `EntityExtractor.java`